modules/apps/frontend-js/frontend-js-components-web/src/main/resources/META-INF/resources/walkthrough/Walkthrough.js

```diff
@@ -0,0 +1,298 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import './Walkthrough.scss';
+
+import ClayButton, {ClayButtonWithIcon} from '@clayui/button';
+import {OverlayMask} from '@clayui/core';
+import {ClayCheckbox} from '@clayui/form';
+import ClayLayout from '@clayui/layout';
+import ClayPopover from '@clayui/popover';
+import {ReactPortal, usePrevious} from '@liferay/frontend-js-react-web';
+import PropTypes from 'prop-types';
+import React, {
+	forwardRef,
+	useCallback,
+	useEffect,
+	useMemo,
+	useRef,
+	useState,
+} from 'react';
+
+import {doAlign} from './doAlign';
+import observeRect from './observeRect';
+
+const Hotspot = forwardRef(({onHotspotClick, trigger}, ref) => {
+	const align = useCallback(() => {
+		if (trigger && ref?.current) {
+			doAlign({
+				points: ['cc', 'tl'],
+				sourceElement: ref.current,
+				targetElement: trigger,
+			});
+		}
+	}, [ref, trigger]);
+
+	useEffect(() => {
+		align();
+	}, [align]);
+
+	useEffect(() => {
+		if (trigger) {
+			return observeRect(trigger, align);
+		}
+	}, [align, trigger]);
+
+	return (
+		<div className="hotspot" onClick={onHotspotClick} ref={ref}>
+			<div className="hotspot-inner" />
+		</div>
+	);
+});
+
+const Walkthrough = ({closeOnClickOutside, closeable, skippable, steps}) => {
+	const [currentStepIndex, setCurrentStepIndex] = useState(0);
+
+	const memoizedTriggerReference = useMemo(() => {
+		const currentNode = steps[currentStepIndex].nodeToHighlight;
+
+		if (currentNode) {
+			return document.querySelector(currentNode);
+		}
+	}, [steps, currentStepIndex]);
+
+	const previousTrigger = usePrevious(memoizedTriggerReference);
```

Comment:
> Can we simplify this by keeping trigger in a state?

React state is meant to be used for things that will trigger a render and I can't see in this flow I created trigger causing more renders

> What is the purpose of memoizedTriggerReference

First, I tried to use a React ref there but weird issues appeared that I don't remember now(8 days ago 😂 ). Then, I reached this following implementation using useMemo with @matuzalemsteles where it looks more stable.

Reply:
> We could then access previous state in the setter, setTrigger((previousTrigger) => {...})

Could you describe more here? 🤔

Reply:
I needed this previousTrigger there because It was used to remove the CSS class for transparent box shadow when the component gets updated

Reply:
Yeah, ideally we don't want a state here because states cause an extra render in the component and we don't want that, so we prefer a `useRef` to stabilize reference during component renders and avoid causing a render if it is reset.

Using `useMemo` is just a move to avoid using state, ref, or `useEffect`, if we use state we would be storing a DOM reference and causing extra rendering whenever we need to update and to update we would need `useEffect`.

The lifecycle of `useEffect` is different from rendering, as in this case, it passes the trigger to the innermost component and it uses in a `useEffect` intern there will be a problem because in some time when trying to get access an element reference he will have an old reference because the `useEffect` call cycle is called from bottom to top instead of top to bottom, also `useEffect` is called after rendering, so `useMemo` in this case helps to stabilize the reference, auto-update and execute during rendering, that is safe because it looks for elements in the DOM that are already rendered and outside React's control otherwise it wouldn't have this information so it would need to use `useRef` and `useEffect` combined and handle the call cycle.